### PR TITLE
Fix broken context extraction for large files.

### PR DESCRIPTION
**Pull request by dmadisetti**
Bug on line 83.

As was, it always returned bool, nil, which would push into the next conditional, which could fall through if the file type is not directly recognized.

In these cases, it's also returning nil. Calling functions treat the response as a table, which leads to unexpected failures.

### CODE REVIEW

File: lua/CopilotChat/context.lua [hunk 1]
```diff
@@ -208,7 +208,8 @@ function M.find_for_query(copilot, opts)
         filetype = filetype,
       })
     else
-      table.insert(outline, M.build_outline(bufnr))
+      -- Since empty just replace with call.
+      outline = M.build_outline(bufnr)
```
Comment (deathbeam) [inline lua/CopilotChat/context.lua:212]: this is incorrect, outline is list of outlines, it do not returns list by itself.

File: lua/CopilotChat/context.lua [hunk 2]
```diff
@@ -154,7 +154,7 @@ function M.build_outline(bufnr)
   get_outline_lines(root)
   local content = table.concat(outline_lines, '\n')
   if content == '' then
-    return
+    return {}
```
Comment (deathbeam) [inline lua/CopilotChat/context.lua:157]: why? Just handle nil outside, do not return empty tables

File: lua/CopilotChat/context.lua [hunk 3]
```diff
@@ -80,12 +80,12 @@ function M.build_outline(bufnr)
...
   local ft = vim.bo[bufnr].filetype
   local lang = vim.treesitter.language.get_lang(ft)
-  local ok, parser = lang and pcall(vim.treesitter.get_parser, bufnr, lang) or false, nil
+  local ok, parser = pcall(vim.treesitter.get_parser, bufnr, lang)
   if not ok or not parser then
     ft = string.gsub(ft, 'react', '')
     ok, parser = pcall(vim.treesitter.get_parser, bufnr, ft)
     if not ok or not parser then
-      return
+      return {}
```
Comment (deathbeam) [inline lua/CopilotChat/context.lua:88]: same here

File: lua/CopilotChat/context.lua [hunk 4]
```diff
@@ -80,12 +80,12 @@ function M.build_outline(bufnr)
   local name = vim.api.nvim_buf_get_name(bufnr)
   local ft = vim.bo[bufnr].filetype
   local lang = vim.treesitter.language.get_lang(ft)
-  local ok, parser = lang and pcall(vim.treesitter.get_parser, bufnr, lang) or false, nil
+  local ok, parser = pcall(vim.treesitter.get_parser, bufnr, lang)
```
Comment (deathbeam) [inline lua/CopilotChat/context.lua:83]: if lang is nil vim.treesitter.get_parser should not be called. if the and/or wasnt working, then just add normal if else around i guess

**Review by deathbeam — CHANGES_REQUESTED** [review]

File: lua/CopilotChat/context.lua (hunk 4 above)
Comment (dmadisetti) [reply lua/CopilotChat/context.lua:83]: from testing, if lang is nil, it's inferred from the buffer. It's non-breaking

but alright

**Comment by dmadisetti** [comment]
Actually tested this time!

File: lua/CopilotChat/context.lua [hunk 5]
```diff
@@ -80,7 +80,10 @@ function M.build_outline(bufnr)
   local name = vim.api.nvim_buf_get_name(bufnr)
   local ft = vim.bo[bufnr].filetype
   local lang = vim.treesitter.language.get_lang(ft)
-  local ok, parser = lang and pcall(vim.treesitter.get_parser, bufnr, lang) or false, nil
+  if not lang then
```
Comment (deathbeam) [inline lua/CopilotChat/context.lua:83]: I did not meant early return but simply to not call the vim.treesitter.get_parser when lang is nil and go straight to fallback. Thats what the original code did (or tried to) e.g avoided the unnecessary call and went to check get_parsr for ft

File: lua/CopilotChat/context.lua [hunk 6]
```diff
@@ -208,11 +211,11 @@ function M.find_for_query(copilot, opts)
         filetype = filetype,
       })
     else
-      table.insert(outline, M.build_outline(bufnr))
+      outline = M.build_outline(bufnr)
```
[the remark this answers is not in the document]
Comment (dmadisetti) [reply lua/CopilotChat/context.lua:214]: That's why my original PR sent back an empty list- with your recommendation you have a mixed type signature

[the remark this answers is not in the document]
Comment (deathbeam) [reply lua/CopilotChat/context.lua:214]: But you cant sometimes send back empty list and sometimes send back just the outline which isnt a list, thats why the original solution did not worked

[the remark this answers is not in the document]
Comment (deathbeam) [reply lua/CopilotChat/context.lua:214]: E.g look at what is build_outline actually returning when it succeeds, I think that is where most of the confusion lies

[the remark this answers is not in the document]
Comment (dmadisetti) [reply lua/CopilotChat/context.lua]: You're right; optionals seem common in Lua. Added a filter to address `nils`

File: lua/CopilotChat/context.lua [hunk 7]
```diff
@@ -80,7 +80,10 @@ function M.build_outline(bufnr)
   local name = vim.api.nvim_buf_get_name(bufnr)
   local ft = vim.bo[bufnr].filetype
   local lang = vim.treesitter.language.get_lang(ft)
-  local ok, parser = lang and pcall(vim.treesitter.get_parser, bufnr, lang) or false, nil
+  ok, parser = false, nil
```
Comment (deathbeam) [inline lua/CopilotChat/context.lua:83]: This should be local

**Review by deathbeam — APPROVED** [review]

**Comment by jellydn** [comment]
@all-contributors add @dmadisetti for code.

**Pull request merged**